### PR TITLE
Implement event filtering for self-profiler.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This is a first sketch for event filtering in the self-profiler, something that we'll want in order to keep profiling overhead low in the common case. The PR contains the commits from https://github.com/rust-lang/rust/pull/59515 and, for the moment, is meant for performance testing.

r? @wesleywiser

### CODE REVIEW

File: src/librustc/util/profiling.rs [hunk 1]
```diff
@@ -27,18 +27,17 @@ pub enum ProfileCategory {
...
+bitflags! {
+    struct EventFilter: u32 {
+        const GENERIC_ACTIVITIES = 0b00000001;
+        const QUERY_PROVIDERS    = 0b00000010;
+        const QUERY_CACHE_HITS   = 0b00000100;
+        const QUERY_BLOCKED      = 0b00001000;
+        const INCR_CACHE_LOADS   = 0b00010000;
+
+        const DEFAULT = Self::GENERIC_ACTIVITIES.bits |
+                        Self::QUERY_PROVIDERS.bits;
```
Comment (wesleywiser) [inline src/librustc/util/profiling.rs]: When parallel rustc becomes the default, adding `Self::QUERY_BLOCKED` might also make sense.

File: src/librustc/util/profiling.rs [hunk 2]
```diff
@@ -55,7 +55,7 @@ pub struct SelfProfiler {
 }
 
 impl SelfProfiler {
-    pub fn new() -> Result<SelfProfiler, Box<dyn Error>> {
+    pub fn new(event_filters: &Option<Vec<String>>) -> Result<SelfProfiler, Box<dyn Error>> {
```
Comment (wesleywiser) [inline src/librustc/util/profiling.rs]: Were you thinking of making [this change](https://rust-lang.zulipchat.com/#narrow/stream/187831-t-compiler.2Fwg-self-profile/topic/event.20filter.20.2F.20level.20of.20detail/near/163101424) in this PR as well?

File: src/librustc/util/profiling.rs (hunk 1 above)
Comment (michaelwoerister) [reply src/librustc/util/profiling.rs]: Yes, I agree.

File: src/librustc/util/profiling.rs (hunk 2 above)
Comment (michaelwoerister) [reply src/librustc/util/profiling.rs]: Eventually, but I'd like to get a perf.rlo run before and after to see if it even makes a difference.

Comment (wesleywiser) [reply src/librustc/util/profiling.rs]: Agreed!

**Comment by wesleywiser** [comment]
@michaelwoerister I rebased your branch on top of #59515 which I've updated to use `measureme@0.2.1`.

@bors try

**Comment by kennytm** [comment]
@rust-timer build 30ce13219eef2589a769b5d694bdc1c0aec0befc

**Comment by wesleywiser** [comment]
Wow, that actually seems pretty good!

**Comment by michaelwoerister** [comment]
Yeah, that looks like a good start.

**Comment by michaelwoerister** [comment]
@rust-timer build 4dd3271ce58c3080ddcc1bf21df704efad2b73c1

**Comment by michaelwoerister** [comment]
That last commit improves performance ever so slightly but I think I'm going to revert it again, for now, because it adds some churn and might regress the case where profiling is disabled. Not quite worth the trouble at this point (also, probably the wrong place to optimize since most of work should happen if something is recorded and not if something is filtered out).

**Comment by michaelwoerister** [comment]
I cleaned up the implementation and added blocking of queries and incr. cache loading to the default event set. Let's see if we can get away with that performance-wise.

@bors try

**Comment by michaelwoerister** [comment]
@rust-timer build 5539376de500270af54c7741ff8075316d950caf

**Comment by michaelwoerister** [comment]
I think this is ready for review. Overhead looks about the same with the new event types added to the default set (slightly worse). Probably still a bit too much to enable this for all perf.rlo runs but we can look into that later.

r? @wesleywiser

**Review by wesleywiser — APPROVED** [review]
looks great to me!

@bors r+

**Comment by wesleywiser** [comment]
That didn't seem to work. I guess bors doesn't see review comments?

@bors r+

File: src/librustc/util/profiling.rs [hunk 3]
```diff
@@ -66,8 +82,38 @@ impl SelfProfiler {
...
+            if unknown_events.len() > 0 {
+                unknown_events.sort();
+                unknown_events.dedup();
+
+                warn!("Unknown self-profiler events specified: {}. Available options are: {}.",
+                    unknown_events.join(", "),
+                    EVENT_FILTERS_BY_NAME.iter()
+                                         .map(|&(name, _)| name.to_string())
+                                         .collect::<Vec<_>>()
+                                         .join(", "));
```
Comment (marieell) [inline src/librustc/util/profiling.rs:108]: I think this warrants being an error instead of just a warning.

**Pull request merged**